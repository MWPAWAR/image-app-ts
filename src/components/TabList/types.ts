export interface TabNavigationPaths {
  RECENTLY_ADDED: string
  FAVORITED: string
}

export type TabNavigationKey = keyof TabNavigationPaths
