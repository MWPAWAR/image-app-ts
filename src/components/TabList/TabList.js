import classNames from 'classnames';
import { useNavigate } from 'react-router-dom';
import { useDispatch } from 'react-redux';
import { resetSelectedImg } from '../../store/imageUiState/imageUiState';
import tabListStyles from './TabList.module.css';

const tabNavigationPaths = {
  RECENTLY_ADDED: 'recently-added',
  FAVORITED: 'favorites',
};

const getCurrentNavigationRoute = () => (
  window.location.pathname.replace('/', '')
);

const TabList = () => {
  const navigate = useNavigate();
  const dispatch = useDispatch();
  const currentRoute = getCurrentNavigationRoute();
  const handleTabClick = (link) => {
    dispatch(resetSelectedImg());
    navigate(`/${tabNavigationPaths[link]}`);
  };

  return (
    <div className={tabListStyles.tabList}>
      <div
        className={classNames(
          tabListStyles.tabListItem,
          currentRoute === tabNavigationPaths.RECENTLY_ADDED ? tabListStyles.active : ''
        )}
        onClick={() => handleTabClick('RECENTLY_ADDED')}
      >
        Recently added
      </div>
      <div
        className={classNames(
          tabListStyles.tabListItem,
          currentRoute === tabNavigationPaths.FAVORITED ? tabListStyles.active : ''
        )}
        onClick={() => handleTabClick('FAVORITED')}
      >
        Favorited
      </div>
    </div>
  );
};

export default TabList;
