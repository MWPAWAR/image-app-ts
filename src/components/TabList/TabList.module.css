.tabList {
  border-bottom: 1px solid #dadce0;
  display: flex;  
  margin: 30px 15px 0 15px;
  width: 90%;
}

.tabListItem {
  cursor: pointer;
  padding-bottom: 15px;
  margin-right: 2%;
  text-align: left;
}

.active {
  color: blue;
  border-bottom: 1px solid blue;
}
