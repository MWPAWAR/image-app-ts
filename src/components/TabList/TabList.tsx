import { FC } from 'react'
import classNames from 'classnames'
import { useNavigate } from 'react-router-dom'
import { useAppDispatch } from '../../customHooks/hooks'
import { resetSelectedImg } from '../../store/imageUiState/imageUiState'
import tabListStyles from './TabList.module.css'
import { TabNavigationKey, TabNavigationPaths } from './types'
import { getString } from '../../i18n'

const tabNavigationPaths: TabNavigationPaths = {
  RECENTLY_ADDED: 'recently-added',
  FAVORITED: 'favorited',
}

const getCurrentNavigationRoute = (): string =>
  window.location.pathname.replace('/', '')

const TabList: FC = () => {
  const navigate = useNavigate()
  const dispatch = useAppDispatch()
  const currentRoute: string = getCurrentNavigationRoute()
  const handleTabClick = (link: string) => {
    dispatch(resetSelectedImg())
    navigate(`/${tabNavigationPaths[link as TabNavigationKey]}`)
  }

  return (
    <div className={tabListStyles.tabList}>
      <div
        className={classNames(
          tabListStyles.tabListItem,
          currentRoute === tabNavigationPaths.RECENTLY_ADDED
            ? tabListStyles.active
            : ''
        )}
        onClick={() => handleTabClick('RECENTLY_ADDED')}
      >
        {getString('tabList.recentlyAdded')}
      </div>
      <div
        className={classNames(
          tabListStyles.tabListItem,
          currentRoute === tabNavigationPaths.FAVORITED
            ? tabListStyles.active
            : ''
        )}
        onClick={() => handleTabClick('FAVORITED')}
      >
        {getString('tabList.favorited')}
      </div>
    </div>
  )
}

export default TabList
