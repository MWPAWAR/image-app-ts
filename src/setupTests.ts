import '@testing-library/jest-dom'

Object.defineProperty(window, 'matchMedia', {
  value: () => {
    return {
      matches: false,
      addListener: () => {},
      removeListener: () => {},
    }
  },
})

export const mockedNavigate = jest.fn()

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedNavigate,
}))
