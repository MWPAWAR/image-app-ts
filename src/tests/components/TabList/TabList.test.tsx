import { MemoryRouter } from 'react-router-dom'
import { Provider } from 'react-redux'
import { render } from '@testing-library/react'
import renderer from 'react-test-renderer'
import userEvent from '@testing-library/user-event'
import configureStore from 'redux-mock-store'
import TabList from '../../../components/TabList/TabList'
import { mockedNavigate } from '../../../setupTests'

describe('TabList', () => {
  const mockStore = configureStore([])
  const store = mockStore({})
  jest.spyOn(store, 'dispatch')

  it('Matches snapshot', () => {
    const tree = renderer.create(
      <MemoryRouter>
        <Provider store={store}>
          <TabList />
        </Provider>
      </MemoryRouter>
    )

    expect(tree).toMatchSnapshot()
  })

  it('calls dispatch if user clicks tabListItem', async () => {
    const { getByText } = render(
      <MemoryRouter>
        <Provider store={store}>
          <TabList />
        </Provider>
      </MemoryRouter>
    )

    jest.spyOn(store, 'dispatch')

    await userEvent.click(getByText('Recently added'))
    expect(store.dispatch).toBeCalled()
    expect(mockedNavigate).toHaveBeenCalledWith('/recently-added')

    await userEvent.click(getByText('Favorited'))
    expect(store.dispatch).toBeCalled()
    expect(mockedNavigate).toHaveBeenCalledWith('/favorited')
  })
})
