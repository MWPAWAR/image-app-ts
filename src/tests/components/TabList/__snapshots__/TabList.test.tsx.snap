// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabList Matches snapshot 1`] = `
<div
  className="tabList"
>
  <div
    className="tabListItem"
    onClick={[Function]}
  >
    Recently added
  </div>
  <div
    className="tabListItem"
    onClick={[Function]}
  >
    Favorited
  </div>
</div>
`;
